Sources/NIO/Codec.swift

```diff
@@ -790,3 +790,270 @@ extension MessageToByteHandler {
+    typealias Decoder = Decoder
+
+    private enum DecodeMode {
+        /// This is a usual decode, ie. not the last chunk
+        case normal
+        /// Last chunk
+        case last
+    }
+
+    private var decoder: Decoder
+    private let maximumBufferSize: Int?
+    internal private(set) var buffer: ByteBuffer?
+
+    /// Initialize a `NIOSingleStepByteToMessageProcessor`.
+    ///
+    /// - parameters:
+    ///     - decoder: The `NIOSingleStepByteToMessageDecoder` to decode the bytes into message.
+    ///     - maximumBufferSize: The maximum number of bytes to aggregate in-memory.
+    public init(_ decoder: Decoder, maximumBufferSize: Int? = nil) {
+        self.decoder = decoder
+        self.maximumBufferSize = maximumBufferSize
+    }
+
+    private func append(_ buffer: ByteBuffer) {
+        if self.buffer == nil {
+            self.buffer = buffer
+        } else {
+            var buffer = buffer
+            self.buffer!.writeBuffer(&buffer)
+        }
+    }
+
+    private func decodeLoop(decodeMode: DecodeMode, seenEOF: Bool = false, _ messageReceiver: (Decoder.InboundOut) -> Void) throws {
+        // we want to call decodeLast once with an empty buffer if we have nothing
+        if decodeMode == .last && (self.buffer == nil || self.buffer!.readableBytes == 0) {
+            var emptyBuffer = self.buffer == nil ? ByteBufferAllocator().buffer(capacity: 0) : self.buffer!
+            let message = try self.decoder.decodeLast(buffer: &emptyBuffer, seenEOF: seenEOF)
+            if message != nil {
+                messageReceiver(message!)
+            }
+            return
+        }
+
+        // buffer can only be nil if we're called from finishProcessing which is handled above
+        assert(self.buffer != nil)
+
+        var decodingState: DecodingState = .continue
+        while self.buffer!.readableBytes > 0 && decodingState == .continue {
+            let message: Decoder.InboundOut?
+            if decodeMode == .normal {
+                message = try self.decoder.decode(buffer: &self.buffer!)
+                if message == nil, let maximumBufferSize = self.maximumBufferSize, self.buffer!.readableBytes > maximumBufferSize {
+                    throw ByteToMessageDecoderError.PayloadTooLargeError()
+                }
+            } else {
+                message = try self.decoder.decodeLast(buffer: &self.buffer!, seenEOF: seenEOF)
+            }
+            if message != nil {
+                messageReceiver(message!)
+            } else {
+                decodingState = .needMoreData
+            }
+            if self.buffer!.readableBytes > 0 {
+                if self.decoder.shouldReclaimBytes(buffer: self.buffer!) {
+                    self.buffer!.discardReadBytes()
+                }
+            }
+        }
+
+        if self.buffer!.readableBytes == 0 {
+            self.buffer!.discardReadBytes()
+        }
+    }
```

Comment:
If we do some careful reordering of the body here we should be able to write this code in such a way that we can use conditional binding to get a non-optional copy of `buffer` instead of having to force unwrap the buffer all the time. This should lead to better code generation because we won't have to add separate branches and `ud2`s for each time we access `self.buffer`.

The trick here will be twofold. Firstly we'll need a local helper function:

```swift
private func withNonCoWBuffer(_ body: (inout ByteBuffer) -> Decoder.InboundOut?) -> Decoder.InboundOut? {
    guard var buffer = self.buffer else {
        return nil
    }

    self.buffer = nil  // To avoid CoW
    let result = body(&buffer)
    self.buffer = buffer
    return result
}
```

Then we can rewrite the function to move all of this logic _except_ the `if message != nil...else...` section into a call to `withNonCoWBuffer`. Then, finally, we can do the `if message != nil` check.

This should remove all the optionality around the buffer, remove the force unwraps, and lead to cleaner looking internal code as well as better code generation.